mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -225,25 +235,456 @@ class PassiveDisplacementCollector : NSObject, PassiveCollector, CLLocationManag
+        #endif
+        
+        recorder.stop() { (action, result, error) in
+            let duration = Date().timeIntervalSince(startTime)
+            
+            if error == nil, let result = result, duration >= 15.0 {
+                #if DEBUG
+                self.debugNotification(title: "Archiving", body: "\(taskState)")
+                #endif
+                
+                taskState.taskResult.appendAsyncResult(with: result)
+                
+                self.archiveManager.archiveAndUpload(taskState) { (_ error: Error?) in
+                    #if DEBUG
+                    self.debugNotification(title: "Done archiving", body: self.archiveErrorDesc(error))
+                    #endif
+                    self.endBackgroundTask()
+                }
+            }
+            else {
+                #if DEBUG
+                if error != nil {
+                    self.debugNotification(title: "Error stopping recorder", body: "\(error!)")
+                }
+                else if result == nil {
+                    self.debugNotification(title: "Stopping recorder", body: "No error reported, but result is nil")
+                }
+                else {
+                    self.debugNotification(title: "Not enough data", body: "only got \(duration) seconds")
+                }
+                #endif
+                taskState.deleteOutputDirectory(error: error)
+                self.endBackgroundTask()
+            }
+        }
+    }
+    
+    // MARK: CLLocationManagerDelegate
+    func locationManagerDidPauseLocationUpdates(_ manager: CLLocationManager) {
+        // This gets called when we are at some place and Location Services thinks we're going to be here
+        // for a while. So, let's set a geofence around it so we'll get pinged when we leave.
+        #if DEBUG
+        debugNotification(title: "Pausing location updates", body: nil)
+        #endif
+        self.locationManagerPaused = true
+        
+        // In order to do that though, we need to get a good fix on our location first, and handle it
+        // in the didUpdateLocations delegate method.
+        self.locationManager?.desiredAccuracy = kCLLocationAccuracyBest
+        self.locationManager?.requestLocation()
+    }
+    
+    func locationManagerDidResumeLocationUpdates(_ manager: CLLocationManager) {
+        #if DEBUG
+        debugNotification(title: "Resuming location updates", body: nil)
+        #endif
+        self.locationManagerPaused = false
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didExitRegion region: CLRegion) {
+        guard region.identifier == kPassiveGaitRegionIdentifier else { return }
+        #if DEBUG
+        debugNotification(title: "Left the region", body: "Starting location updates")
+        #endif
+
+        self.locationManager?.stopMonitoring(for: region)
+        self.locationManager!.desiredAccuracy = kCLLocationAccuracyKilometer
```

Comment:
Um, probably better to use `let locationManager = self.locationManager` above or else to use a consistent `?`

Reply:
oops